File: algorithms/07_typescript/data_structures/dsu.ts
export class DSU {
  private parent: number[];
  private size: number[];

  constructor(n: number) {
    this.parent = new Array(n);
    this.size = new Array(n).fill(1);
    for (let i = 0; i < n; i++) this.parent[i] = i;
  }

  find(x: number): number {
    if (this.parent[x] !== x) this.parent[x] = this.find(this.parent[x]);
    return this.parent[x];
  }

  union(a: number, b: number): boolean {
    let ra = this.find(a), rb = this.find(b);
    if (ra === rb) return false;
    if (this.size[ra] < this.size[rb]) [ra, rb] = [rb, ra];
    this.parent[rb] = ra;
    this.size[ra] += this.size[rb];
    return true;
  }

  connected(a: number, b: number): boolean {
    return this.find(a) === this.find(b);
  }
}

const dsu = new DSU(5);
dsu.union(0, 1);
dsu.union(1, 2);
if (!dsu.connected(0, 2) || dsu.connected(0, 3)) throw new Error("DSU failed");
console.log("[TypeScript DSU] Union-find with path compression verified");

File: algorithms/21_scala/data_structures/DSU.scala
object DSU {
  class UnionFind(n: Int) {
    private val parent = (0 until n).toArray
    private val rank = Array.fill(n)(0)

    def find(x: Int): Int = {
      if (parent(x) != x) parent(x) = find(parent(x))
      parent(x)
    }

    def union(x: Int, y: Int): Unit = {
      val rx = find(x)
      val ry = find(y)
      if (rx == ry) return
      if (rank(rx) < rank(ry)) parent(rx) = ry
      else if (rank(rx) > rank(ry)) parent(ry) = rx
      else { parent(ry) = rx; rank(rx) += 1 }
    }

    def connected(x: Int, y: Int): Boolean = find(x) == find(y)
  }

  def main(args: Array[String]): Unit = {
    val dsu = new UnionFind(5)
    dsu.union(0, 1)
    dsu.union(1, 2)
    dsu.union(3, 4)
    assert(dsu.connected(0, 2), "union failed")
    assert(!dsu.connected(0, 3), "spurious union")
    println("[Scala DSU] Disjoint set union verified")
  }
}
